### PR TITLE
feat: close position button

### PR DESCRIPTION
**Pull request by chrisling-dev**
https://www.notion.so/prepo/Draft-Tickets-Board-8d821ff8adac4efbb71cd1878d3d4c88?p=deccc77df5f249a1bce6c062aed3cb53&pm=s

### CODE REVIEW

**Review by chrisling-dev — COMMENTED** [review]
self reviewed

File: apps/frontend/core-dapp/src/features/trade/close-trade/PositionsSlideUp.tsx [hunk 1]
```diff
@@ -37,15 +37,18 @@ const PositionsSlideUp: React.FC = () => {
...
   const { connected } = web3Store
   const { userPositions } = portfolioStore
-  const { slideUpContent, setSlideUpContent, selectedPosition } = tradeStore
+  const { closingPosition, slideUpContent, setSlideUpContent, selectedPosition } = tradeStore
 
   const hasNoPosition = userPositions !== undefined && userPositions.length === 0
   // slide up is disabled if:
   // - not connected: button shows "Connect Wallet"
   // - loading posiitions: loading skeleton in button
   // - user has no position and nothing selected: Show "No Opened Position"
```
Comment (XavierEkkel) [inline apps/frontend/core-dapp/src/features/trade/close-trade/PositionsSlideUp.tsx:46]: nit: update comment to reflect `closingPosition` state

File: apps/frontend/core-dapp/src/features/trade/close-trade/PositionsSlideUp.tsx [hunk 2]
```diff
@@ -37,15 +37,18 @@ const PositionsSlideUp: React.FC = () => {
   const { portfolioStore, tradeStore, web3Store } = useRootStore()
   const { connected } = web3Store
   const { userPositions } = portfolioStore
-  const { slideUpContent, setSlideUpContent, selectedPosition } = tradeStore
+  const { closingPosition, slideUpContent, setSlideUpContent, selectedPosition } = tradeStore
 
   const hasNoPosition = userPositions !== undefined && userPositions.length === 0
   // slide up is disabled if:
   // - not connected: button shows "Connect Wallet"
   // - loading posiitions: loading skeleton in button
```
Comment (XavierEkkel) [inline apps/frontend/core-dapp/src/features/trade/close-trade/PositionsSlideUp.tsx:45]: nit:
```suggestion
  // - loading positions: loading skeleton in button
```

File: apps/frontend/core-dapp/src/features/trade/TradeStore.ts [hunk 3]
```diff
@@ -295,6 +369,16 @@ export class TradeStore {
...
+  async approveClosePositions(): Promise<void> {
+    if (!this.selectedPosition || this.approvingClosePositions) return
+    this.approvingClosePositions = true
+    await this.selectedPosition.token.unlockPermanently('UNISWAP_SWAP_ROUTER')
+    runInAction(() => {
+      this.approvingClosePositions = false
+    })
+  }
+
+  // delete this function when we remove ClosePositionSummary
```
Comment (XavierEkkel) [inline apps/frontend/core-dapp/src/features/trade/TradeStore.ts:381]: nit: use 'TODO' in these comments so they're easy to find later
```suggestion
  // TODO: delete this function when we remove ClosePositionSummary
```

File: apps/frontend/core-dapp/src/features/trade/TradeStore.ts [hunk 4]
```diff
@@ -170,8 +173,79 @@ export class TradeStore {
...
+      !this.selectedPosition ||
+      this.selectedPosition.totalValueBN === undefined ||
+      this.selectedPosition.priceBN === undefined ||
+      this.closePositionValueBN === undefined
+    )
+      return undefined
+
+    // we converted price to priceBN with ERC20_UNITS to hold decimals in the price for more precise calculation (18 decimals)
+    // so when converting from value to long/short token amount, we need to multiply the denominator
+    // because when we divide priceBN later, the price has been multiplied by the same priceDenominator
```
Comment (XavierEkkel) [inline apps/frontend/core-dapp/src/features/trade/TradeStore.ts:231]: nit: specify where 'later' is exactly
and consider moving price denominator out into a shared `WEI_DENOMINATOR` constant across files

**Review by XavierEkkel — APPROVED** [review]
LGTM after nits

**Pull request merged**